src/background.c

```diff
@@ -262,6 +262,122 @@ handle_notify_background (XdpImplBackground *object,
   return TRUE;
 }
 
+/* This is conservative, but lets us avoid escaping most
+   regular Exec= lines, which is nice as that can sometimes
+   cause problems for apps launching desktop files. */
+static gboolean
+need_quotes (const char *str)
+{
+  const char *p;
+
+  for (p = str; *p; p++)
+    {
+      if (!g_ascii_isalnum (*p) &&
+          strchr ("-_%.=:/@", *p) == NULL)
+        return TRUE;
+    }
+
+  return FALSE;
+}
+
+static char *
+maybe_quote (const char *str)
+{
+  if (need_quotes (str))
+    return g_shell_quote (str);
+  return g_strdup (str);
+}
+
+static gboolean
+handle_enable_autostart (XdpImplBackground *object,
+                         GDBusMethodInvocation *invocation,
+                         const char *arg_app_id,
+                         gboolean    arg_enable,
+                         const char * const *arg_commandline,
+                         gboolean    arg_activatable)
+{
+  gboolean result = FALSE;
+  g_autofree char *dir = NULL;
+  g_autofree char *file = NULL;
+  g_autofree char *path = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GString) args = NULL;
+  g_autoptr(GKeyFile) keyfile = NULL;
+
+  g_debug ("background: handle EnableAutostart");
+
+  file = g_strconcat (arg_app_id, ".desktop", NULL);
+  dir = g_build_filename (g_get_user_config_dir (), "autostart", NULL);
+  path = g_build_filename (dir, file, NULL);
+
+  if (!arg_enable)
+    {
+      unlink (path);
+      g_debug ("Removed %s", path);
+      goto out;
+    }
+
+  if (g_mkdir_with_parents (dir, 0755) != 0)
+    {
+      g_warning ("Failed to create dirs %s", dir);
+      goto out;
+    }
+
+  args = g_string_new ("");
+  g_string_append (args, "flatpak run");
+  if (arg_commandline && arg_commandline[0])
```

Comment:
I think xdg-desktop-portal should generate the full commandline, because if the calling app is not a flatpak (like, maybe its not contained at all) it should not generate a flatpak run command, and the backend implementation should not have to care about this.